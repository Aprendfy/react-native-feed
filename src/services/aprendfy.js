const BASE_URL = 'https://aprendfy.herokuapp.com/v1/app/';
const CATEGORIES = 'categories/';
const POSTS = 'posts/';

type categoriesResponsePayloadType = {
    _id: string,
    __v: number,
    name: string,
    color: string,
    created_at: Date,
    updated_at: Date
}

type authorType = {
  _id: string,
  name: string
}

type postsPayloadResponseType = {
  _id: string,
  title: string,
  category: string,
  readingTime: string,
  level: string,
  body: string,
  image: string,
  author: authorType
}

const fetchGetParams = {
  method: 'get',
  headers: {
    /* eslint max-len: ["error", 500] */
    Authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1OTYzNDljNDZjZDQyNjAwMTE2Y2U0NTciLCJ1cGRhdGVkX2F0IjoiMjAxNy0wNy0xMFQwOTozMjo1Mi4yNTFaIiwiY3JlYXRlZF9hdCI6IjIwMTctMDctMTBUMDk6MzI6NTIuMjUxWiIsIm5hbWUiOiJVc3XDoXJpbyBQYWRyw6NvIiwiZW1haWwiOiJ1c3VhcmlvQHBhZHJhby5jb20iLCJfX3YiOjAsInVzZXJTdGF0dXMiOiJQRU5ESU5HIiwidHlwZSI6IlBVQkxJU0hFUiIsImlhdCI6MTQ5OTk0Njk4OX0.aQKSVDOSDni0OqIrCDugmbSKNSs_V5zsnxkqlaQoMWo',
    'Content-Type': 'application/json'
  }
}

export async function getFeedsByCategory(category) {
  try {
    const response = await fetch(`${BASE_URL}${POSTS}?category=${category}`, fetchGetParams);

    if (!response.ok) {
      throw Error(response);
    }

    const responseJson = await response.json();
    return mapPostsResponsePayloadIntoArray(responseJson.payload, category);
  } catch (error) {
    console.error(error);
    return [];
  }
}

export async function getAllCategories() {
  try {
    const response = await fetch(`${BASE_URL}${CATEGORIES}`, fetchGetParams);

    if (!response.ok) {
      throw Error(response);
    }

    const responseJson = await response.json();
    return mapCategoriesResponsePayloadIntoArray(responseJson.payload);
  } catch (error) {
    console.error(error);
    return [];
  }
}

const mapCategoriesResponsePayloadIntoArray = (payloadResponse: categoriesResponsePayloadType) => {
  return payloadResponse.map(category => toCategoryRepresentation(category));
};

const mapPostsResponsePayloadIntoArray = (payloadReponse: postsPayloadResponseType, category) => {
  return {[category]: payloadReponse};
}

const toCategoryRepresentation = (obj) => {
  return { title: obj.name, color: obj.color };
};
