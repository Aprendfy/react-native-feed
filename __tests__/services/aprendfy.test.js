import FetchMock from 'fetch-mock';
import { getAllCategories, getFeedsByCategory } from '../../src/services/aprendfy';
import CATEGORIES_FIXTURE from '../assets/fixtures/getCategoriesResponse.json';
import POST_FIXTURE from '../assets/fixtures/getPostsResponse.json';

describe('Aprendfy backend service', () => {
  it('Should retrieve all categories', async () => {
    FetchMock.get('https://aprendfy.herokuapp.com/v1/app/categories', CATEGORIES_FIXTURE);

    const categories = await getAllCategories();

    expect(categories).toHaveLength(6);
    categories.forEach(category => expect(Object.keys(category)).toEqual(['title', 'color']));
  });

  it('Should retrieve all posts by category', async () => {
    const desiredCategory = 'facebook';
    FetchMock.get(`https://aprendfy.herokuapp.com/v1/app/posts/?category=${desiredCategory}`, POST_FIXTURE);

    const posts = await getFeedsByCategory(desiredCategory);

    expect(posts).toHaveProperty(desiredCategory);
    expect(posts[desiredCategory]).toHaveLength(2);
    expect(posts[desiredCategory][0]).toMatchSnapshot();
  });
});
