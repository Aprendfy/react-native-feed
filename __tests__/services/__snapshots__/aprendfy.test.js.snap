// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aprendfy backend service Should retrieve all posts by category 1`] = `
Object {
  "author": "Usuário Padrão",
  "body": "Ainda assim, existem dúvidas a respeito de como a",
  "category": "Facebook",
  "image": "http://i.huffpost.com/gen/3971736/images/o-HAPPY-PEOPLE-facebook.jpg",
  "level": "Iniciante",
  "readingTime": "4 minutos",
  "title": "Criar uma Página no Facebook",
}
`;
